Use code generation for joins with filter function

losipiuk opened the pull request:
This PR extends the non-equality-outer-joins. It adds flow for execution joins with filter function using compiled version of PagesHashStrategy.

martint commented:
@erichwang, can you review?

erichwang commented on presto-main/src/main/java/com/facebook/presto/sql/gen/JoinCompiler.java:
```diff
@@ -243,6 +254,22 @@ private static void generateConstructor(ClassDefinition classDefinition,
+
+        constructor.comment("check if join filter function is passed only if field is generated");
+        if (joinFilterFunctionFieldOptional.isPresent()) {
+            constructor.append(invokeStatic(Preconditions.class, "checkArgument", void.class, ImmutableList.of(boolean.class, Object.class),
```
it seems weird that we are adding checkArgs into the compiled code

losipiuk replied:
Yeah - somewhat.
We generate the constructor. But the call is elsewhere. If this check is not present the code will fail when generated `PagesHashStrategy.applyFilterFunction` is called.
I can drop it if you like. Please confirm this or that way.

erichwang commented on presto-main/src/main/java/com/facebook/presto/operator/PagesIndex.java:
```diff
@@ -315,8 +315,8 @@ public PagesHashStrategy createPagesHashStrategy(List<Integer> joinChannels, Opt
         try {
-            return joinCompiler.compilePagesHashStrategyFactory(types, joinChannels)
-                    .createPagesHashStrategy(ImmutableList.copyOf(channels), hashChannel);
+            return joinCompiler.compilePagesHashStrategyFactory(types, joinChannels, joinFilterFunction.map(f -> f.getClass()))
```
dont abbreviate

erichwang replied:
also, instead of passing around an Optional<Class>, it seems like the only thing you need here is a boolean indicating whether a filter function exists?

losipiuk replied:
Yeah - I passed the JFF class as one of the code versions declared the `joinFilterFunction` field in generated code as that class and not as just `JoinFilterFunction`. But it did not have any impact on performance so I dropped that.

I will replace Optional with boolean flag.

haozhun replied:
`JoinFilterFunction::getClass`

IntelliJ should warn you on this line: "can be replaced with method reference". If it doesn't, please turn the warning on.

haozhun replied:
Issue 1:

This function has only one callsite, and `joinFilterFunction` is always `Optional.empty()`. Therefore, inline this function into `createPagesHashStrategy(List<Integer> joinChannels, Optional<Integer> hashChannel)`.

Issue 2:
* For compiled `LookupSource`, we delegate to `JoinFilterFunction` directly. In generated `PagesHashStrategy`, `hasFilterFunction` always return false and `applyFilterFunction` always returns true.
  * But if you look at `JoinCompiler.generateXxxFilterFunction`, this not apparent at all.
* For interpreted `InMemoryLookupSource`, we delegate to `SimplePagesHashStrategy.applyFilterFunction`, which delegates to `JoinFilterFunction`.

This is confusing and hard to reason about because there are 2 possible places that have the logic to evaluate JFF. And you always have it in exactly one of the 2 places, either PagesHashStrategy or InMemoryLookupSource. Let's remove JFF from PagesHashStrategy entirely and have `InMemoryLookupSource` call `JoinFilterFunction` directly.

erichwang commented on presto-main/src/main/java/com/facebook/presto/sql/gen/JoinCompiler.java:
```diff
@@ -237,6 +241,41 @@ private static void generateConstructor(ClassDefinition classDefinition,
+            BytecodeExpression firstChannel = thisVariable.getField(channelFields.get(0));
+            BytecodeBlock loopBody = new BytecodeBlock();
+            constructor.append(
+                    new ForLoop()
```
can you add a comment as to the java equivalent of what this is doing?

losipiuk replied:
done

erichwang commented on presto-main/src/main/java/com/facebook/presto/sql/gen/ExpressionCompiler.java:
```diff
@@ -210,4 +256,55 @@ public String toString()
+
+        public JoinFilterCacheKey(RowExpression filter, int leftBlocksSize)
+        {
+            this.filter = filter;
```
require non null

losipiuk replied:
done

erichwang commented on presto-main/src/main/java/com/facebook/presto/sql/gen/ExpressionCompiler.java:
```diff
@@ -210,4 +256,55 @@ public String toString()
+        public JoinFilterCacheKey(RowExpression filter, int leftBlocksSize)
+        {
+            this.filter = filter;
+            this.leftBlocksSize = leftBlocksSize;
```
checkArg that this is >= 0

losipiuk replied:
This should be true now as we do not use compiled join-filter functions this for CROSS JOIN with condition. But it is not strong requirement for this class so I do not feel that we want to enforce it here.

erichwang commented:
@martint, can you take a look at this too?

losipiuk commented:
@martint , @erichwang comments addressed

haozhun commented on presto-main/src/main/java/com/facebook/presto/sql/gen/JoinCompiler.java:
```diff
@@ -635,16 +634,16 @@ private static void generatePositionEqualsPositionMethod(
         getFilterFunctionMethod.getBody()
-                .append(invokeStatic(Optional.class, "empty", Optional.class))
-                .ret(Optional.class);
+                .append(constantBoolean(false))
```
`.append(constantFalse().ret())`

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/PagesHashStrategy.java:
```diff
@@ -93,9 +91,9 @@
 
     /**
-     * Returns filter function assigned to this PagesHashStrategy.
+     * Returns if PagesHashStrategy has filter function assigned.
```
nit: `@return if PagesHashStrategy has filter function assigned`

haozhun commented on presto-main/src/main/java/com/facebook/presto/operator/PagesIndex.java:
```diff
@@ -338,12 +338,13 @@ public LookupSource createLookupSource(List<Integer> joinChannels, Optional<Inte
 
             try {
-                LookupSourceFactory lookupSourceFactory = joinCompiler.compileLookupSourceFactory(types, joinChannels);
+                LookupSourceFactory lookupSourceFactory = joinCompiler.compileLookupSourceFactory(types, joinChannels, filterFunction.map(f -> f.getClass()));
```
`JoinFilterFunction::getClass`

losipiuk replied:
Cool it actually works. I used lambda because intelliJ complained here with "type mismatch" error. But javac compiles it correctly.

haozhun commented:
@losipiuk Make sure you see this comment: https://github.com/prestodb/presto/pull/5810#discussion_r75229014. It's automatically hidden because the line is overwritten in a later commit.

@martint, @erichwang and I had a further discussion about where JoinFilterFunction should be put. It appears like it would make sense to put it in LookupJoinOperator (instead of PagesHashStrategy or LookupSource). We always instantiate `LookupJoinOperator` through `JoinProbeCompiler.internalCompileJoinOperatorFactory`. Therefore, having a virtual callsite in `LookupJoinOperator` should be ok. This way, it will benefit `InMemoryJoinHash`, generated `LookupSource`, and `IndexLookupSource` in one shot.

haozhun commented:
I'm removing myself from the reviewers for now because @erichwang originally wanted my help to look at code generation in this pull request. But it appears like this pull request requires quite some additional work on the other parts first. @erichwang can always add me back as necessary sometime later.

erichwang commented:
After thinking about this a bit more, I think actually the LookupSource is the right place to do the filtering. In an ideal world for me, the filtering would happen in the LookupJoinOperator, but to do so would require that we have some way to lift the values out of the LookupSource and ProbeSource to do some filtering. In lieu of that, I think the current approach is the next best option.

The main thing that sticks out to me here is that the PagesHashStrategy compiles an optional filter function in and we have to deal with the optional even after the code is compiled (which seems silly). What do you guys think about having the filter function just default to a result of true if it does not exist? Then we can get rid of the "hasFilterFunction" method and just always assume that we have something that works. Any concerns on the performance impact here?

Finally, a lot of work has been done here for the InMemoryHashJoin path of the code. We should eventually get some of this work added to the IndexJoin path which also uses the LookupSource. Based on my investigation, I think we can do this easily by adjusting the implementations of IndexedData to also take the filter function and passing the probe data to it.

losipiuk commented:
> After thinking about this a bit more, I think actually the LookupSource is the right place to do the filtering. In an ideal world for me, the filtering would happen in the LookupJoinOperator, but to do so would require that we have some way to lift the values out of the LookupSource and ProbeSource to do some filtering. In lieu of that, I think the current approach is the next best option.

Some time have passed since we implemented initial non-compiled join-filter-function support, but I recall exactly same thinking process. Initially we wanted to have the logic in LookupJoinOperator - but decided in putting it in LookupSource for exact same reason you stated. 
 
> The main thing that sticks out to me here is that the PagesHashStrategy compiles an optional filter function in and we have to deal with the optional even after the code is compiled (which seems silly). What do you guys think about having the filter function just default to a result of true if it does not exist? Then we can get rid of the "hasFilterFunction" method and just always assume that we have something that works. Any concerns on the performance impact here?

Performance was our concern here. We wanted to minimise the impact of handling filter function for cases without it, which are common ones at least for now.
The `applyFilterFunction` current looks like this:
```java
   private boolean applyFilterFilterFunction(int leftPosition, int rightPosition, Block[] rightBlocks)
    {
        if (!filterFunctionPresent) {
            return true;
        }

        long pageAddress = addresses.getLong(leftPosition);
        int blockIndex = decodeSliceIndex(pageAddress);
        int blockPosition = decodePosition(pageAddress);

        return pagesHashStrategy.applyFilterFunction(blockIndex, blockPosition, rightPosition, rightBlocks);
    }
```
The question is if calculating `blockIndex` and `blockPosition` is negligible. 
Actually we can move the `decode*` calls into generated `PHS.applyFilterFunction` method. And leave just `long pageAddress = addresses.getLong(leftPosition);` 
in `InMemoryJoinHash.applyFilterFilterFunction`. Maybe cost of calling it is fine as we are dropping `if(!filterFunctionPresent)` statement.

I am fine with micro benchmarking that. But it would be nice to not treat that as merge-blocker for this PR and handle that as separate work.
 
> Finally, a lot of work has been done here for the InMemoryHashJoin path of the code. We should eventually get some of this work added to the IndexJoin path which also uses the LookupSource. Based on my investigation, I think we can do this easily by adjusting the implementations of IndexedData to also take the filter function and passing the probe data to it.

Yeah it seems so. We did not play with that (honestly mostly because we do not have usecases for index joins - as no public connector supports them).
I can try to work something out outside of this PR. But I will have hard time to do benchmarking of that for example - so I will need your help on that.

losipiuk commented:
I pushed new with haozhun's comments applied. With squashed fixup comments.

erichwang commented:
@losipiuk, I'm still having a bit of a disconnect with the Optional. I took another look at the code and was wondering what value we get by putting the applyFilterFunction in the pages hash strategy? It seems like the only thing we are doing in that class is to just to forward the call to another method? In that case, it would be just as efficient to not generate new bytecode. Instead, we should then just pass the Optional filter directly to the lookup source and have it deal with that class directly. If that works, that should simplify quite a lot of this code and the only thing we would need to generate byte code for is the filter expression.

losipiuk commented:
@erichwang. But then you would need to expose build side  channels out of the PHS. Do LookupSource can pass them to JFF. Currently PHS does not have methods for that and that would kinda break its abstraction.

Or maybe you are suggesting to pass build channels to JFF during constructon and have then both in JFF and in PHS. That could work for Hash join but it does not seem play well with index join.

The latter solution should work without changes in PHS. But it seems that complexity will just be moved to join filter function compiler. I may miss sth as I do not have computer at hand to verify the details.

The former one still requires some changes to PHS to build channelArrays list. And expose it to lookup source.

erichwang commented:
@losipiuk, yes, the PHS currently contains all of the build data, but that is not something that needs to be the rule. In fact, one of the long term refactorings that we need to do is to try to break the PHS into smaller pieces. Given that we don't consider PHS to be the sole source of build data, the JFF has little value being compiled into the PHS.

> Or maybe you are suggesting to pass build channels to JFF during constructon and have then both in JFF and in PHS. That could work for Hash join but it does not seem play well with index join.

This is close, but what I am suggesting here is to just pass the build data and JFF directly to the InMemoryJoinHash and have that class deal with it directly. The only thing that needs to be byte code generated here is the JFF. Another option is to wrap up the build data and JFF into a single object (plain class, not byte code generated) to be passed to the InMemoryJoinHash, and in that case we would need to make sure that container class is isolated with IsolatedClass to avoid having that method become polymorphic.

losipiuk commented:
I will implement version with `JoinFunctionVerifier` which wraps JFF and build channels.

losipiuk commented:
@erichwang Take a look at refactored code. 
I also spotted that JoinProbe was not generated when filter function was in use - the 3b279e5 fixes that

erichwang commented on presto-main/src/main/java/com/facebook/presto/sql/gen/JoinFilterFunctionCompiler.java:
```diff
@@ -0,0 +1,190 @@
+        BytecodeNode visitorBody = filter.accept(visitor, scope);
+
+        Variable result = scope.declareVariable(boolean.class, "result");
+        body.append(visitorBody)
```
use com.facebook.presto.bytecode.expression.BytecodeExpressions#inlineIf instead of a separate variable

erichwang commented on presto-main/src/main/java/com/facebook/presto/sql/gen/JoinFilterFunctionCompiler.java:
```diff
@@ -0,0 +1,190 @@
+                .append(new IfStatement()
+                        .condition(wasNullVariable)
+                        .ifTrue(constantFalse().ret())
+                        .ifFalse(result.ret()));
```
looks like result is never initialized?

losipiuk replied:
It is initialized.
`visitorBody` leaves computation result on stack and then it is put into `result` variable via `.putVariable(result)`.

losipiuk replied:
I tried to change it to code without `result` variable
```java
        body.append(visitorBody)
                .append(new IfStatement()
                        .condition(wasNullVariable)
                        .ifTrue().pop().push(false)
                )
                .retBoolean();
```
but for some reason it does not work.

dain replied:
I'd do this:
```java
        // evaluate filter and assign to result variable
        Variable result = scope.declareVariable(boolean.class, "result");
        body.append(filter.accept(visitor, scope))
                .putVariable(result);

        // return (wasNull ? false : result);
        body.append(inlineIf(wasNullVariable, constantFalse(), result).ret());

erichwang commented on presto-main/src/main/java/com/facebook/presto/sql/gen/JoinFilterFunctionCompiler.java:
```diff
@@ -0,0 +1,190 @@
+        generateConstructor(classDefinition, cachedInstanceBinder);
+    }
+
+    private void generateFilterMethod(ClassDefinition classDefinition, CallSiteBinder callSiteBinder, CachedInstanceBinder cachedInstanceBinder, RowExpression filter, int leftBlocksSize)
```
@dain, can you review this?

erichwang commented on presto-main/src/main/java/com/facebook/presto/operator/EmptyJoinFilterFunctionVerifier.java:
```diff
@@ -0,0 +1,26 @@
+
+import com.facebook.presto.spi.block.Block;
+
+public class EmptyJoinFilterFunctionVerifier
```
instead of newing this up everywhere, lets just create a public static instance here

losipiuk replied:
done

losipiuk commented:
@erichwang I adressed your last comment

erichwang commented:
thanks @losipiuk, let's get @dain to do one more pass over this

dain commented:
@losipiuk if you didn't notice, the travis tests are failing for this one

dain commented on presto-main/src/main/java/com/facebook/presto/operator/StandardJoinFilterFunctionVerifier.java:
```diff
@@ -0,0 +1,70 @@
+        return channelArrays;
+    }
+
+    public boolean applyFilterFunction(int leftBlockIndex, int leftPosition, int rightPosition, Block[] allRightBlocks)
```
Missing `@Override`

losipiuk replied:
done

dain commented on presto-main/src/main/java/com/facebook/presto/operator/StandardJoinFilterFunctionVerifier.java:
```diff
@@ -0,0 +1,70 @@
+        this.filterFunction = requireNonNull(filterFunction, "filterFunction can not be null");
+    }
+
+    private List<Block[]> buildChannelArrays(List<List<Block>> channels)
```
I'm not sure why this is a separate method.  The I would just inline this into the constructor.

dain replied:
I find this code quite confusing.  I think it is taking a column oriented data set (i.e., a list of columns each of which is a list of blocks) and converting it into a page oriented data (i.e., a list of pages each of which is a list of column-blocks), but the name and variables seem to indicate the conversion is in the other direction.

As mentioned above, I would inline this into the constructor, and then I would rename as follows:
`channelsArray` -> `pages` (both variable and field)
`i` -> `pageIndex`
`j` -> `channelIndex`

losipiuk replied:
Makes sense. Done.

dain commented on presto-main/src/main/java/com/facebook/presto/operator/StandardJoinFilterFunctionVerifier.java:
```diff
@@ -0,0 +1,70 @@
+    private List<Block[]> buildChannelArrays(List<List<Block>> channels)
+    {
+        if (channels.isEmpty()) {
+            return new ArrayList<>();
```
This is assigned directly to the `channelArrays` field, so this should be an `ImmutableMap`

losipiuk replied:
Done

dain commented on presto-main/src/main/java/com/facebook/presto/operator/StandardJoinFilterFunctionVerifier.java:
```diff
@@ -0,0 +1,70 @@
+
+    public StandardJoinFilterFunctionVerifier(JoinFilterFunction filterFunction, List<List<Block>> channels)
+    {
+        this.channelArrays = buildChannelArrays(requireNonNull(channels, "channels can not be null"));
```
For consistency, the field, parameters and assignments should generally be in the same order.  In this case, this assignment should be below `filterFunction`

losipiuk replied:
done

dain commented on presto-main/src/main/java/com/facebook/presto/operator/StandardJoinFilterFunctionVerifier.java:
```diff
@@ -0,0 +1,70 @@
+        }
+
+        int pagesCount = channels.get(0).size();
+        List<Block[]> channelArrays = new ArrayList<>();
```
Use `ImmutableList.builder()`

losipiuk replied:
Done

dain commented on presto-main/src/main/java/com/facebook/presto/operator/PagesIndex.java:
```diff
@@ -343,7 +353,8 @@ public LookupSource createLookupSource(List<Integer> joinChannels, Optional<Inte
                         ImmutableList.<List<Block>>copyOf(channels),
-                        hashChannel);
+                        hashChannel,
+                        createJoinFilterFunctionVerifier(filterFunction, ImmutableList.<List<Block>>copyOf(channels)));
```
The `<List<Block>>` is not necessary in Java8.  You IDE should be warning you about this.

losipiuk replied:
Yeah. This is also true for `copyOf` 2 lines above this one ;). For some reason IntelliJ adds those unnecessary specifications sometimes and does not emit warning. 
Removed both.

dain commented on presto-main/src/main/java/com/facebook/presto/operator/PagesIndex.java:
```diff
@@ -326,6 +326,16 @@ public PagesHashStrategy createPagesHashStrategy(List<Integer> joinChannels, Opt
+    private JoinFilterFunctionVerifier createJoinFilterFunctionVerifier(Optional<JoinFilterFunction> filterFunction, List<List<Block>> channels)
+    {
+        if (filterFunction.isPresent()) {
+            return new StandardJoinFilterFunctionVerifier(filterFunction.get(), channels);
```
Remove the channels parameter and just make this line:
```java
            return new StandardJoinFilterFunctionVerifier(filterFunction.get(), ImmutableList.copyOf(channels));
```

dain commented on presto-main/src/main/java/com/facebook/presto/operator/PagesIndex.java:
```diff
@@ -355,12 +366,12 @@ public LookupSource createLookupSource(List<Integer> joinChannels, Optional<Inte
         PagesHashStrategy hashStrategy = new SimplePagesHashStrategy(
                 types,
-                ImmutableList.<List<Block>>copyOf(channels),
+                ImmutableList.copyOf(this.channels),
```
this is redundant here

losipiuk replied:
done

dain commented on presto-main/src/main/java/com/facebook/presto/operator/InMemoryJoinHash.java:
```diff
@@ -168,7 +167,7 @@ public final long getNextJoinPosition(long currentJoinPosition, int probePositio
     {
         long currentJoinPosition = startJoinPosition;
-        while (filterFunctionPresent && currentJoinPosition != -1 && !applyFilterFilterFunction(Ints.checkedCast(currentJoinPosition), probePosition, allProbeChannelsPage.getBlocks())) {
+        while (currentJoinPosition != -1 && !applyFilterFilterFunction(Ints.checkedCast(currentJoinPosition), probePosition, allProbeChannelsPage)) {
```
Instead of doing these multiple checkedCasts, I would do it once like this:
```java
        int currentJoinPosition = Ints.checkedCast(startJoinPosition);
        while (currentJoinPosition != -1 && !applyFilterFilterFunction(currentJoinPosition, probePosition, allProbeChannelsPage)) {
            currentJoinPosition = positionLinks[currentJoinPosition];
        }
        return currentJoinPosition;

losipiuk replied:
done

dain commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -1491,9 +1491,23 @@ private LookupSourceSupplier createLookupJoinSource(
             Optional<Integer> buildHashChannel = buildHashSymbol.map(channelGetter(buildSource));
 
             OperatorFactory operatorFactory;
```
this variable should be declared below the new block

losipiuk replied:
done

dain commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -1491,9 +1491,23 @@ private LookupSourceSupplier createLookupJoinSource(
+                        metadata,
+                        sqlParser,
+                        sourceTypes,
+                        rewrittenFilter);
```
This needs to be updated for the prepared statement changes

dain commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -1491,9 +1491,23 @@ private LookupSourceSupplier createLookupJoinSource(
                 Map<Symbol, Integer> joinSourcesLayout = createJoinSourcesLayout(buildSource.getLayout(), probeLayout);
-                return new InterpretedFilterFunction(filter, context.getTypes(), joinSourcesLayout, metadata, sqlParser, context.getSession());
+                Map<Integer, Type> sourceTypes = new HashMap<>();
+                joinSourcesLayout.forEach((symbol, field) -> sourceTypes.put(field, context.getTypes().get(symbol)));
```
This is simpler:
```java
                Map<Integer, Type> sourceTypes = joinSourcesLayout.entrySet().stream()
                        .collect(toImmutableMap(Entry::getValue, entry -> context.getTypes().get(entry.getKey()))));

losipiuk replied:
Hmm. If you say so :). Changed.

dain commented on presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java:
```diff
@@ -1491,9 +1491,23 @@ private LookupSourceSupplier createLookupJoinSource(
             OperatorFactory operatorFactory;
-            Optional<JoinFilterFunction> filterFunction = node.getFilter().map(filter -> {
+
+            Optional<JoinFilterFunction> filterFunction = node.getFilter().map(filterExpression -> {
```
I'd pull this out into a separate method like this:
```java
            Optional<JoinFilterFunction> filterFunction = node.getFilter()
                    .map(filterExpression -> compileJoinFilterFunction(filterExpression, probeLayout, buildSource.getLayout(), context.getTypes(), context.getSession()));

        private JoinFilterFunction compileJoinFilterFunction(
                Expression filterExpression,
                Map<Symbol, Integer> probeLayout,
                Map<Symbol, Integer> buildLayout,
                Map<Symbol, Type> types,
                Session session)
        {
            Map<Symbol, Integer> joinSourcesLayout = createJoinSourcesLayout(buildLayout, probeLayout);

            Map<Integer, Type> sourceTypes = joinSourcesLayout.entrySet().stream()
                    .collect(toImmutableMap(Entry::getValue, entry -> types.get(entry.getKey())));

            SymbolToInputRewriter symbolToInputRewriter = new SymbolToInputRewriter(joinSourcesLayout);
            Expression rewrittenFilter = ExpressionTreeRewriter.rewriteWith(symbolToInputRewriter, filterExpression);

            IdentityHashMap<Expression, Type> expressionTypes = getExpressionTypesFromInput(
                    session,
                    metadata,
                    sqlParser,
                    sourceTypes,
                    rewrittenFilter);

            RowExpression translatedFilter = toRowExpression(rewrittenFilter, expressionTypes);
            return compiler.compileJoinFilterFunction(translatedFilter, buildLayout.size()).get();
        }

losipiuk replied:
done

dain commented on presto-main/src/main/java/com/facebook/presto/sql/gen/JoinFilterFunctionCompiler.java:
```diff
@@ -0,0 +1,190 @@
+    private void generateFilterMethod(ClassDefinition classDefinition, CallSiteBinder callSiteBinder, CachedInstanceBinder cachedInstanceBinder, RowExpression filter, int leftBlocksSize)
+    {
+        // todo handle TRY expression
+        Map<CallExpression, MethodDefinition> tryMethodMap = ImmutableMap.of();
```
What happens if there is a try expression?

losipiuk replied:
Damn - totally forgot about this todo. It will probably fail in some ugly way. I will work on that.

dain commented on presto-main/src/main/java/com/facebook/presto/sql/gen/JoinFilterFunctionCompiler.java:
```diff
@@ -0,0 +1,190 @@
+        // todo handle TRY expression
+        Map<CallExpression, MethodDefinition> tryMethodMap = ImmutableMap.of();
+
+        //int leftPosition, Block[] leftBlocks, int rightPosition, Block[] rightBlocks
```
space after `//`

dain commented on presto-main/src/main/java/com/facebook/presto/sql/gen/JoinFilterFunctionCompiler.java:
```diff
@@ -0,0 +1,190 @@
+        };
+    }
+
+    private void generateConstructor(ClassDefinition classDefinition, CachedInstanceBinder cachedInstanceBinder)
```
This method can be static.  The `cachedInstanceBinder` parameter is never used.  The `constructor` variable is not needed.  Your IDE should warn you about all of these.

dain replied:
This method is not needed.  Simply replace caller with `classDefinition.declareDefaultConstructor(a(PUBLIC));`

losipiuk replied:
> Your IDE should warn you about all of these

And it does. 
Sorry about that. To little care here.

dain commented on presto-main/src/main/java/com/facebook/presto/sql/gen/JoinFilterFunctionCompiler.java:
```diff
@@ -0,0 +1,190 @@
+            final int leftBlocksSize,
+            final Variable wasNullVariable)
+    {
+        return new RowExpressionVisitor<Scope, BytecodeNode>()
```
I'd like to see this and PageProcessor

The new class would be
```java
import com.facebook.presto.bytecode.BytecodeNode;
import com.facebook.presto.bytecode.Scope;
import com.facebook.presto.bytecode.Variable;
import com.facebook.presto.bytecode.control.IfStatement;
import com.facebook.presto.bytecode.expression.BytecodeExpression;
import com.facebook.presto.spi.type.Type;
import com.facebook.presto.sql.relational.CallExpression;
import com.facebook.presto.sql.relational.ConstantExpression;
import com.facebook.presto.sql.relational.InputReferenceExpression;
import com.facebook.presto.sql.relational.RowExpressionVisitor;
import com.google.common.primitives.Primitives;
import io.airlift.slice.Slice;

import java.util.function.BiFunction;

import static com.facebook.presto.sql.gen.SqlTypeBytecodeExpression.constantType;
import static java.util.Objects.requireNonNull;

class InputReferenceCompiler
        implements RowExpressionVisitor<Scope, BytecodeNode>
{
    private final BiFunction<Scope, Integer, BytecodeExpression> blockResolver;
    private final BiFunction<Scope, Integer, BytecodeExpression> positionResolver;
    private final Variable wasNullVariable;
    private final CallSiteBinder callSiteBinder;

    public InputReferenceCompiler(
            BiFunction<Scope, Integer, BytecodeExpression> blockResolver,
            BiFunction<Scope, Integer, BytecodeExpression> positionResolver,
            Variable wasNullVariable,
            CallSiteBinder callSiteBinder)
    {
        this.blockResolver = requireNonNull(blockResolver, "blockResolver is null");
        this.positionResolver = requireNonNull(positionResolver, "positionResolver is null");
        this.wasNullVariable = requireNonNull(wasNullVariable, "wasNullVariable is null");
        this.callSiteBinder = requireNonNull(callSiteBinder, "callSiteBinder is null");
    }

    @Override
    public BytecodeNode visitInputReference(InputReferenceExpression node, Scope scope)
    {
        int field = node.getField();
        Type type = node.getType();

        BytecodeExpression block = blockResolver.apply(scope, field);
        BytecodeExpression position = positionResolver.apply(scope, field);

        Class<?> javaType = type.getJavaType();
        if (!javaType.isPrimitive() && javaType != Slice.class) {
            javaType = Object.class;
        }

        IfStatement ifStatement = new IfStatement();
        ifStatement.condition(block.invoke("isNull", boolean.class, position));

        ifStatement.ifTrue()
                .putVariable(wasNullVariable, true)
                .pushJavaDefault(javaType);

        String methodName = "get" + Primitives.wrap(javaType).getSimpleName();
        ifStatement.ifFalse(constantType(callSiteBinder, type).invoke(methodName, javaType, block, position));

        return ifStatement;
    }

    @Override
    public BytecodeNode visitCall(CallExpression call, Scope scope)
    {
        throw new UnsupportedOperationException("not yet implemented");
    }

    @Override
    public BytecodeNode visitConstant(ConstantExpression literal, Scope scope)
    {
        throw new UnsupportedOperationException("not yet implemented");
    }
}
``` 

And the usage in PageProcessor would be:
```java
    private static RowExpressionVisitor<Scope, BytecodeNode> fieldReferenceCompiler(final CallSiteBinder callSiteBinder, final Variable positionVariable, final Variable wasNullVariable)
    {
        return new InputReferenceCompiler(
                (scope, field) -> scope.getVariable("block_" + field),
                (scope, field) -> positionVariable,
                wasNullVariable,
                callSiteBinder);
    }
```

And the usage in this class would be this variable declaration (it is simpler than a method for a single call):
```java
        RowExpressionVisitor<Scope, BytecodeNode> fieldReferenceCompiler = new InputReferenceCompiler(
                (scope, field) -> field < leftBlocksSize ? leftBlocks.getElement(field) : rightBlocks.getElement(field - leftBlocksSize),
                (scope, field) -> field < leftBlocksSize ? leftPosition : rightPosition,
                wasNullVariable,
                callSiteBinder);
```

losipiuk replied:
Nice :)
I put it as separate commit and set you as an author.
I can squash it to my commit if you prefer.

dain commented on presto-main/src/main/java/com/facebook/presto/sql/gen/ExpressionCompiler.java:
```diff
@@ -42,33 +43,66 @@
+                public Class<? extends JoinFilterFunction> load(JoinFilterCacheKey key)
+                        throws Exception
+                {
+                    ClassDefinition classDefinition = new ClassDefinition(
```
Move this code to a method, say `compileJoinFilterInternal`, so the cache field declaration is small (like the ones above)

losipiuk replied:
done

dain commented on presto-main/src/main/java/com/facebook/presto/sql/gen/ExpressionCompiler.java:
```diff
@@ -42,33 +43,66 @@
 
-    private final LoadingCache<CacheKey, Class<? extends PageProcessor>> pageProcessors = CacheBuilder.newBuilder().maximumSize(1000).build(
-            new CacheLoader<CacheKey, Class<? extends PageProcessor>>()
+    private final LoadingCache<ProcessorCacheKey, Class<? extends PageProcessor>> pageProcessors = CacheBuilder.newBuilder().maximumSize(1000).build(
```
Looking at the changes in this file.  The only method reused from the existing code is `generateToString` (and that reuse isn't necessary).  So, instead of adding more stuff to this class, move all of this new stuff to `JoinFilterFunctionCompiler` and make it injectable, so it can be bound directly into `LocalExecutionPlanner`.

You can replace the call to `generateToString` with this:
```java
        classDefinition.declareMethod(a(PUBLIC), "toString", type(String.class))
                .getBody()
                .append(constantString(toStringHelper(classDefinition.getType().getJavaClassName())
                        .add("filter", filter)
                        .add("leftBlocksSize", leftBlocksSize)
                        .toString()).ret());

dain commented on presto-main/src/main/java/com/facebook/presto/sql/gen/JoinCompiler.java:
```diff
@@ -103,6 +121,26 @@ public LookupSourceFactory compileLookupSourceFactory(List<? extends Type> types
         }
     }
 
+    public JoinFilterFunctionVerifierFactory compileJoinFilterFunctionVerifierFactory(Optional<JoinFilterFunction> joinFilterFunction)
```
I'm curious why we have `JoinFilterFunction`abstraction at all if we really want a `JoinFilterFunctionVerifier`.  The we could simply generate a `JoinFilterFunctionVerifier`, which is basically the existing code with on extra `leftPages` field, and this method:
```java
    public boolean applyFilterFunction(int leftBlockIndex, int leftPosition, int rightPosition, Block[] allRightBlocks)
    {
        return filterFunction.filter(leftPosition, leftPages.get(leftBlockIndex), rightPosition, allRightBlocks);
    }

losipiuk replied:
I think the problem with that approach is that currently `filterFunction` instance is constructed in `LocalExecutionPlanner.createLookupJoinSource`. And we do not have the build pages available at this point. So we cannot just replace `filterFunction` with `filterFunctionVerifier` instance here.
We need to either introduce `JoinFilterFunctionVerifierFactory` or pass build pages as parameter on each call to applyFilter.
I am not super convinced that would be strictly better.

Anyway. I will apply other comments first and leave this as separate final refactoring.

dain commented:
Assign this back to me, when the changes are applied.

losipiuk commented:
@dain 
I applied the comments. One thing I skipped Is merging JoinFilterFunction and JoinFilterFunctionVerifier. See https://github.com/prestodb/presto/pull/5810#discussion_r77822122.
I may try to merge those and see how it would look like if you insist.

I squashed smaller changes to original commits. 
I left some changes as separate commits for clarity. Those can be squashed before merge:
* Support for function requiring `ConnectorSession` parameter:  https://github.com/prestodb/presto/pull/5810/commits/f9c7e2d
* Support for `TRY`: https://github.com/prestodb/presto/pull/5810/commits/37e36a5
* Reuse code for fieldReferenceCompiler: https://github.com/prestodb/presto/pull/5810/commits/d710fe8 - Here I set you as author.

There is also small cleanup commit:  https://github.com/prestodb/presto/pull/5810/commits/16cdc77

dain commented:
@losipiuk there is a test failure with index joins:
```
testHighCardinalityReducedIndexProbeKey(com.facebook.presto.tests.TestLocalQueriesIndexed)  Time elapsed: 0.301 sec  <<< FAILURE! java.lang.NullPointerException: null
	at com.facebook.presto.operator.InMemoryJoinHash.applyFilterFilterFunction(InMemoryJoinHash.java:230)
	at com.facebook.presto.operator.InMemoryJoinHash.getNextJoinPositionFrom(InMemoryJoinHash.java:170)
	at com.facebook.presto.operator.InMemoryJoinHash.getNextJoinPosition(InMemoryJoinHash.java:164)
	at com.facebook.presto.operator.index.IndexSnapshot.getNextJoinPosition(IndexSnapshot.java:55)

dain commented:
BTW the rest of the PR looks good, so just throw the fix on the end and I'll get this merged.

losipiuk commented:
@dain 
I fixed that with 3761bce727dab6a0086be79c92bdb69a839afcc4 and rebased to master.
Please squash before merging.

losipiuk commented:
@dain TL;DR It seems that we need a `filterFunctionPresent` back.

I did final round of micro benchmarks and it does not look that great after refactorings.
Previously I compared the speed of code with and without refactorings for cases when filter function is actually used. And that case was fine. Now I tested the common case when there are only equality join conjuncts and no filter function is needed.

I tested with query:
`SELECT COUNT(*) FROM t t1 LEFT JOIN t t2 ON t1.a = t2.a`
for table generated as:
`CREATE TABLE t AS SELECT orderkey % 1000 as a, orderkey % 1000 as b FROM tpch.sf1.orders LIMIT 10000`

Baseline is code where I changed
`InMemoryJoinHash.getNextJoinPositionFrom(int startJoinPosition, int probePosition, Page allProbeChannelsPage)` so it just returns `startJoinPosition` without ever calling `applyJoinFilterFuntion`:
```
BenchmarkNonEquiOuterJoins.benchmark  avgt   50  10.528 ± 0.391  ms/op
```

Current state of code is much worse:
```
BenchmarkNonEquiOuterJoins.benchmark  avgt   50  13.579 ± 0.814  ms/op
```

You can make it a bit faster by passing right blocks as `Page` instead of `Block[]` to `JoinFilterFunctionVerifier`.
The result then is:
```
BenchmarkNonEquiOuterJoins.benchmark  avgt   50  12.589 ± 0.494  ms/op
```

I managed to get major speedup by precalculating `filterFunctionPresent` in constructor as `!(filterFunctionVerifier instanceof EmptyJoinFilterFunctionVerifier);`
And changing `getNextJoinPosition` so while condition uses it:
```java
 while (filterFunctionPresent && currentJoinPosition != -1 && !applyFilterFilterFunction(currentJoinPosition, probePosition, allProbeChannelsPage))
```
With such change I get reasonable results:
```
 BenchmarkNonEquiOuterJoins.benchmark  avgt   50  10.953 ± 0.430  ms/op
```

So it seems to me that we should get rid of `EmptyJoinFilterFunctionVerifier`. And instead pass `Optional<JoinFilterFunctionVerifier` to `InMemoryJoinHash`. The `JFFV` would no longer be interface but concrete class with contents of  current `StandardJoinFilterFunctionVerifier`

What do you think?

fmeiser commented:
I think it would be possible that the filter function is only push down to JoinProbe which is compiled anyway? You loose some time with the overhead from the call to advance but your interfaces are cleaner.

losipiuk commented:
>I think it would be possible that the filter function is only push down to JoinProbe which is compiled anyway? You loose some time with the overhead from the call to advance but your interfaces are cleaner.

@Fmeiser  are you suggesting to move `JoinFilterFunction` verification to `JoinProbe`? 

This could work (if I am not missing something) but I would rather keep it in `InMemoryJoinHash`. 
You have multiple implementations of `JoinProbe` and you would have to modify all of them. Also that involves changing generated code which is not super nice. 
And one of the reasons of introducing `JoinFilterVerifier` here is to limit the amount of code which is generated. 
Evaluating it in JoinProbe could potentially be faster as you could skip `JFF` evaluation at all if there is no filter function involved. Yet `filterFunctionPresent` also does the job. Though I agree it is not super elegant.

fmeiser commented:
@losipiuk Yes.
It must not be modified all implementations, when you provide a class `FilteringJoinProbe`. Then the `JoinProbeFactory ` can decide whether it implements the filter with a `FilteringJoinProbe`, the implementation supports a filter or compile the filter into the `JoinProbe`.
The generated code should consist of not more than one `if `and some variables to call `FilterFunction`?

losipiuk commented:
Well @fmeiser. I don't know. It seems that what you say should work.
Though I am not 100% sure that it may be much nicer.
You would need to make changes to:
- `LookupJoinOperator.joinCurrentPosition` so the call to `lookupSource.getNextJoinPosition(joinPosition, probe.getPosition(), probe.getPage());` is proxied via JoinProbe.
- we would need to pass build pages to `JoinProbeFactory` and currently `LookupJoinOperator`has no direct access to those.
- The changes to generated `JoinProbe` would probably be not that huge, yet still.

I may be biased here a bit as I have limited resources to work on that :).
I would rather merge the good-enough solution. There is always place for improvements.

@dain your call

dain commented:
@losipiuk I would have `InMemoryJoinHash` take `Optional<JoinFilterFunctionVerifier>` and then remove `EmptyJoinFilterFunctionVerifier`.  In the constructor unwrap `Optional` into a `JoinFilterFunctionVerifier` and do a null check in the loop.  Add a comment that we unwrap for performance reasons.

losipiuk commented:
@Dain. It should be good to go. Squashed and rebased. 
I restarted travis build as one of the jobs failed. It looked unrelated. 
Let's wait until it completes.

dain commented:
Merged, thanks!
